File: android/features/simulator/src/main/java/com/bolsaterminal/features/simulator/TaxCalculatorViewModel.kt
package com.bolsaterminal.features.simulator

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.bolsaterminal.domain.tax.IrpfCalculator
import com.bolsaterminal.domain.tax.IrpfResult
import com.bolsaterminal.domain.tax.TaxCurrency
import com.bolsaterminal.domain.tax.TaxLotEntry
import com.bolsaterminal.domain.tax.TaxSaleEntry
import com.bolsaterminal.domain.tax.TaxTradeResult
import dagger.hilt.android.lifecycle.HiltViewModel
import java.time.LocalDate
import javax.inject.Inject

@HiltViewModel
class TaxCalculatorViewModel @Inject constructor() : ViewModel() {

    val lots = mutableStateListOf<TaxLotEntry>()
    val sales = mutableStateListOf<TaxSaleEntry>()
    var carryForwardText by mutableStateOf("0")

    val tradeResults: List<TaxTradeResult>
        get() = sales.map { IrpfCalculator.matchFIFO(lots, it) }

    val irpf: IrpfResult
        get() = IrpfCalculator.buildResult(tradeResults, carryForwardText.toDoubleOrNull() ?: 0.0)

    fun addLot(symbol: String, buyDate: LocalDate, buyPrice: Double, quantity: Double, currency: TaxCurrency, eurRate: Double) {
        lots.add(TaxLotEntry(symbol = symbol.uppercase(), buyDate = buyDate, buyPrice = buyPrice, quantity = quantity, currency = currency, eurRateAtBuy = if (eurRate > 0) eurRate else 1.08))
    }

    fun addSale(symbol: String, sellDate: LocalDate, sellPrice: Double, quantity: Double, currency: TaxCurrency, eurRate: Double) {
        sales.add(TaxSaleEntry(symbol = symbol.uppercase(), sellDate = sellDate, sellPrice = sellPrice, quantity = quantity, currency = currency, eurRateAtSell = if (eurRate > 0) eurRate else 1.08))
    }

    fun removeLot(id: String) {
        lots.removeAll { it.id == id }
    }

    fun removeSale(id: String) {
        sales.removeAll { it.id == id }
    }
}

File: macos/BolsaTerminal/Features/Simulator/TaxCalculatorViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class TaxCalculatorViewModel {
    var lots: [TaxLotEntry] = []
    var sales: [TaxSaleEntry] = []
    var carryForward: String = "0"

    var tradeResults: [TaxTradeResultCalc] {
        sales.map { IrpfCalculator.matchFIFO(lots: lots, sale: $0) }
    }

    var irpf: IrpfResultCalc {
        IrpfCalculator.buildResult(trades: tradeResults, carryForwardLoss: Double(carryForward) ?? 0)
    }

    func addLot(symbol: String, buyDate: Date, buyPrice: Double, quantity: Double, currency: TaxCurrency, eurRateAtBuy: Double) {
        lots.append(TaxLotEntry(
            symbol: symbol.uppercased(), buyDate: buyDate, buyPrice: buyPrice, quantity: quantity,
            currency: currency, eurRateAtBuy: eurRateAtBuy > 0 ? eurRateAtBuy : 1.08
        ))
    }

    func addSale(symbol: String, sellDate: Date, sellPrice: Double, quantity: Double, currency: TaxCurrency, eurRateAtSell: Double) {
        sales.append(TaxSaleEntry(
            symbol: symbol.uppercased(), sellDate: sellDate, sellPrice: sellPrice, quantity: quantity,
            currency: currency, eurRateAtSell: eurRateAtSell > 0 ? eurRateAtSell : 1.08
        ))
    }

    func removeLot(_ id: TaxLotEntry.ID) {
        lots.removeAll { $0.id == id }
    }

    func removeSale(_ id: TaxSaleEntry.ID) {
        sales.removeAll { $0.id == id }
    }
}
